### PR TITLE
Add scatter gather sample

### LINKED ISSUE #6: Create a sample for Scatter-Gather pattern
Create a sample for [Scatter-Gather](https://www.enterpriseintegrationpatterns.com/patterns/messaging/BroadcastAggregate.html) pattern

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
Fixes #6

### CODE REVIEW

File: scatter-gather/scatter-gather.bal [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import ballerina/http;
+
+service /api/v1 on new http:Listener(8080) {
+    
+    resource function get vehicles(string dropOffDateTime, string dropOffLocation, string pickupDateTime, string pickupLocation) returns json|error {
+        return check requestRentalCar(dropOffDateTime, dropOffLocation, pickupDateTime, pickupLocation);
+    }
+}
+
+function transformResponse(map<anydata|error> responses) returns json|error {
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]: Let's break this to two functions `transformAvis` and `transformHotwire`. Then we can call it inside the each worker.

File: scatter-gather/scatter-gather.bal [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    json[] results = [];
+    foreach var [_, value] in responses.entries() {
+        if value is json {
+            foreach json item in <json[]>(check value.result) {
+                results.push({
+                    carType: check item.carType,
+                    price: check item.price,
+                    resultLink: check item.resultLink
+                });
+            }
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]: See if you can use Ballerina query syntax.

File: scatter-gather/scatter-gather.bal [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        } else if value is xml {
+            foreach xml item in value/<Response>/<Car> {
+                results.push({
+                    carType: (item/<CarType>).data(),
+                    price: (item/<Price>).data(),
+                    resultLink: (item/<ResultLink>).data()
+                });
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]: Same here, let's try query syntax.

File: scatter-gather/scatter-gather.bal [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            }
+        }
+    }
+    return results;
+}
+
+function requestRentalCar(string dropOffDateTime, string dropOffLocation, string pickupDateTime, string pickupLocation) returns json|error {
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]: Let's replace this `json` with a type, such as `CarRental`

File: scatter-gather/tests/case-01.http [hunk 5]
```diff
@@ -0,0 +1,2 @@
+
```
Comment (manuranga) [inline scatter-gather/tests/case-01.http]: empty line.

**Review by manuranga — CHANGES_REQUESTED** [review]
Please address the review comments.

File: scatter-gather/scatter-gather.bal [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import ballerina/http;
+
+service /api/v1 on new http:Listener(8080) {
+    
+    resource function get vehicles(string dropOffDateTime, string dropOffLocation, string pickupDateTime, string pickupLocation) returns json|error {
+        return check requestRentalCar(dropOffDateTime, dropOffLocation, pickupDateTime, pickupLocation);
```
Comment (DimuthuMadushan) [inline scatter-gather/scatter-gather.bal]:
```suggestion
        return requestRentalCar(dropOffDateTime, dropOffLocation, pickupDateTime, pickupLocation);
```

File: scatter-gather/scatter-gather.bal (hunk 4 above)
Comment (ushirask) [inline scatter-gather/scatter-gather.bal]: Shall we move this function to immediate below of the service

File: scatter-gather/scatter-gather.bal [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    CarRental[] result = [];
+    map<CarRental[]|error> mapResult = wait {hotwire, avis};
+    from CarRental[]|error carRental in mapResult
+        where carRental is CarRental[]
+        do {
+            result.push(...carRental);
+        };
+    return result;
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]: Let's `return wait {hotwire, avis}` and change the return type accordingly. This is to reduce the length of the sample.

File: scatter-gather/scatter-gather.bal [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+|};
+
+service /api/v1 on new http:Listener(8080) {
+    
+    resource function get vehicles(string dropOffDateTime, string dropOffLocation, string pickupDateTime, string pickupLocation) returns CarRental[]|error {
+        return requestRentalCar(dropOffDateTime, dropOffLocation, pickupDateTime, pickupLocation);
+    }
+}
+
+function requestRentalCar(string dropOffDateTime, string dropOffLocation, string pickupDateTime, string pickupLocation) returns CarRental[]|error {
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]: inline this function.

File: scatter-gather/scatter-gather.bal [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    map<CarRental[]|error> mapResult = wait {hotwire, avis};
+    from CarRental[]|error carRental in mapResult
+        where carRental is CarRental[]
+        do {
+            result.push(...carRental);
+        };
+    return result;
+}
+
+function transformHotwireResponse(xml response) returns CarRental[]|error =>
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]: Can we not return an error form this?

**Review by manuranga — CHANGES_REQUESTED** [review]
Please address the review comments.

File: scatter-gather/scatter-gather.bal (hunk 7 above)
Comment (LakshanWeerasinghe) [reply scatter-gather/scatter-gather.bal]: Even we return the `wait {hotwire, avis}`, later we have to process to update the final response correctly.

File: scatter-gather/scatter-gather.bal (hunk 8 above)
Comment (LakshanWeerasinghe) [reply scatter-gather/scatter-gather.bal]: Do you mean to have this body inside the resource function?

File: scatter-gather/scatter-gather.bal (hunk 9 above)
Comment (LakshanWeerasinghe) [reply scatter-gather/scatter-gather.bal]: We can.

File: scatter-gather/scatter-gather.bal (hunk 8 above)
Comment (manuranga) [reply scatter-gather/scatter-gather.bal]: yes.

File: scatter-gather/scatter-gather.bal (hunk 7 above)
Comment (manuranga) [reply scatter-gather/scatter-gather.bal]: I don't think scatter gather patterns says you have to pick a single response, replying all in a map is still valid.

Comment (LakshanWeerasinghe) [reply scatter-gather/scatter-gather.bal]: Your argument is valid. But when a developer runs this sample he should see something nice. Having a array is better than having a map which contains arrays as values.

Comment (manuranga) [reply scatter-gather/scatter-gather.bal]: We are on a strict budget for number of lines we can use, which is a higher priority. It's ok if we can write this concisely. Can we use `map.values()` ?
btw why are arrays better then maps?

Comment (LakshanWeerasinghe) [reply scatter-gather/scatter-gather.bal]: Here the map output will be in this structure `{"avis": CarRental[], "hotwire": CarRental[] }`. This way I said returning an array would be better.

Comment (manuranga) [reply scatter-gather/scatter-gather.bal]: Is that structure bad?

Comment (LakshanWeerasinghe) [reply scatter-gather/scatter-gather.bal]: In a real scenario the client will only interested in the array result. If use the `map.values()` the result will be a nested array.

File: scatter-gather/scatter-gather.bal [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+|};
+
+service /api/v1 on new http:Listener(8080) {
+    
+    resource function get vehicles(string dropOffDateTime, string dropOffLocation, string pickupDateTime, string pickupLocation) returns CarRental[]|error {
+        return requestRentalCar(dropOffDateTime, dropOffLocation, pickupDateTime, pickupLocation);
+    }
+}
+
+function requestRentalCar(string dropOffDateTime, string dropOffLocation, string pickupDateTime, string pickupLocation) returns CarRental[]|error {
```
Comment (anupama-pathirage) [inline scatter-gather/scatter-gather.bal]: Break line at max of 120 chars.

Comment (anupama-pathirage) [reply scatter-gather/scatter-gather.bal]: Check all places.

File: scatter-gather/scatter-gather.bal (hunk 7 above)
Comment (LakshanWeerasinghe) [reply scatter-gather/scatter-gather.bal]: If we go with the current approach following will be the output.
```json
[
  {
    "carType": "Chevy Uplander",
    "price": "USD 229.84",
    "resultLink": "http://siva.com/quote?id=95900"
  },
  {
    "carType": "Hyundai Accent",
    "price": "USD 214.40",
    "resultLink": "http://hotline.com/quote/1234"
  },
  {
    "carType": "Ford Freestar",
    "price": "USD 242.13",
    "resultLink": "http://hotline.com/quote/1257"
  }
]
```

Comment (LakshanWeerasinghe) [reply scatter-gather/scatter-gather.bal]: Response after your suggestion, 
```json
{
  "siva": [
    {
      "carType": "Chevy Uplander",
      "price": "USD 229.84",
      "resultLink": "http://siva.com/quote?id=95900"
    }
  ],
  "hotline": [
    {
      "carType": "Hyundai Accent",
      "price": "USD 214.40",
      "resultLink": "http://hotline.com/quote/1234"
    },
    {
      "carType": "Ford Freestar",
      "price": "USD 242.13",
      "resultLink": "http://hotline.com/quote/1257"
    }
  ]
}
```

File: scatter-gather/scatter-gather.bal (hunk 10 above)
Comment (LakshanWeerasinghe) [reply scatter-gather/scatter-gather.bal]: Fixed.

File: scatter-gather/scatter-gather.bal [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+            country_code="US", dropOff_date=dropOffDateTime, dropoff_location=dropOffLocation,
+            pickup_date=pickupDateTime, pickup_location=pickupLocation);
+            return avisResponse.result.'map(r => {carType: r.carType, price: r.price, resultLink: r.resultLink});
+        }
+
+        return map from CarRental[]|error carRental in wait {hotwire, avis} 
+                where carRental is CarRental[] && carRental.length() > 0
+                let string vendor = (<regexp:Span>(<regexp:Groups[]> re `/(\w+)\.`
+                    .findAllGroups(carRental[0].resultLink))[0][1]).substring()
+                select [vendor, carRental];
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]: Can we do something simpler eg:
```
var r = check wait {hotwire, avis} ;
return [...check r.hotwire, ...check r.avis];
```

File: scatter-gather/scatter-gather.bal [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    }
+}
+
+function transformHotwireResponse(xml response) returns CarRental[] =>
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]: let's use curly

File: scatter-gather/scatter-gather.bal (hunk 11 above)
Comment (LakshanWeerasinghe) [reply scatter-gather/scatter-gather.bal]: We cannot use this approach. If we want to get rid of the error part we need to iterate through the responses.

File: scatter-gather/scatter-gather.bal [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+            AvisResponse avisResponse = check avisEP -> /cars/catalog/v1/vehicles/rates.get(brand="Avis", 
+            country_code="US", dropOff_date=dropOffDateTime, dropoff_location=dropOffLocation,
+            pickup_date=pickupDateTime, pickup_location=pickupLocation);
+            return transformAvisResponse(avisResponse);
+        }
+
+        return map from CarRental[]|error carRental in wait {hotwire, avis} 
+                where carRental is CarRental[] && carRental.length() > 0
+                let string vendor = carRental[0].vendor
+                select [vendor, carRental];
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]:
```
var all = wait {hotwire, avis};
return map from var [vendor, carRental] in all.entries()
           where vendor !is error
           select [vendor, carRental];
```

File: scatter-gather/scatter-gather.bal [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+        };
+
+function transformAvisResponse(AvisResponse response) returns CarRental[] {
+    return from var item in response.result
+        select {
+            carType: item.carType,
+            price: item.price,
+            resultLink: item.resultLink,
+            vendor: item.vendor
+        };
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]:
```
return from var {carType, price, resultLink, vendor} in response.result
           select {carType, price, resultLink, vendor};
```

File: scatter-gather/scatter-gather.bal [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+service /api/v1 on new http:Listener(8080) {
+    
+    resource function get vehicles(string dropOffDateTime, string dropOffLocation, string pickupDateTime,
+    string pickupLocation) returns map<CarRental[]>|error {
+    
+        worker hotwire returns CarRental[]|error {
+            http:Client hotwireEP = check new ("http://api.hotwire.com.balmock.io");
+            http:Response hotwireResponse = check hotwireEP -> /v1/search/car.get(pickup=pickupLocation, 
+            dest=dropOffLocation, startDate=pickupDateTime, endDate=dropOffDateTime);
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]: indent.

File: scatter-gather/scatter-gather.bal [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+            return transformHotwireResponse(check hotwireResponse.getXmlPayload());
+        }
+
+        worker avis returns CarRental[]|error {
+            http:Client avisEP = check new ("http://api.avis.com.balmock.io");
+            AvisResponse avisResponse = check avisEP -> /cars/catalog/v1/vehicles/rates.get(brand="Avis", 
+            country_code="US", dropOff_date=dropOffDateTime, dropoff_location=dropOffLocation,
+            pickup_date=pickupDateTime, pickup_location=pickupLocation);
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]: indent.

File: scatter-gather/scatter-gather.bal [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+        worker avis returns CarRental[]|error {
+            AvisResponse avisResponse = check avisEP -> /cars/catalog/v1/vehicles/rates.get(brand="Avis", 
+                country_code="US", dropOff_date=dropOffDateTime, dropoff_location=dropOffLocation,
+                pickup_date=pickupDateTime, pickup_location=pickupLocation);
+            return transformAvisResponse(avisResponse);
+        }
+
+        var responses = wait {hotwire, avis} ;
+        return map from var [vendor, carRental] in responses.entries() 
+                where carRental is CarRental[]
```
Comment (manuranga) [inline scatter-gather/scatter-gather.bal]: let's say `!is error`

**Review by manuranga — APPROVED** [review]

**Pull request merged**